### PR TITLE
ENH: Add support for reading Neuralynx data

### PR DESCRIPTION
**Pull request by KristijanArmeni**
#### Reference issue
Work in progress for #11874

#### What does this implement/fix?
- Implements a draft of `RawNeuralynx()` class, subclassing `BaseRaw()`, and the `read_raw_neuralynx()` wrapper.
- relies on functionalities in the [`NeuralynxIO`](https://neo.readthedocs.io/en/latest/iolist.html#neo.io.NeuralynxIO) class from the Neo package for reading header information

#### To-Do

- [ ] **Testing data.** We are missing testing data to test this implementation and move forward. I can ask if there's ways to get a sample recording on our end. Can also ask on forums (MNE, Fieldtrip etc.) for any samples.

### CODE REVIEW

File: mne/io/neuralynx/__init__.py [hunk 1]
```diff
@@ -0,0 +1 @@
+from neuralynx import read_raw_neuralynx
```
Comment (larsoner) [inline mne/io/neuralynx/__init__.py:1]:
```suggestion
from .neuralynx import read_raw_neuralynx
```

File: mne/io/neuralynx/neuralynx.py [hunk 2]
```diff
@@ -0,0 +1,85 @@
+import os.path as op
+from mne.io import BaseRaw
+from mne.io.utils import _read_segments_file
+from mne import create_info
+from mne.utils import logger, verbose, fill_doc
```
Comment (larsoner) [inline mne/io/neuralynx/neuralynx.py:5]: All imports in `mne/` to other MNE stuff should be relative (other than tests, which should use absolute imports):
```suggestion
from ..base import BaseRaw
from ..utils import _read_segments_file, _mult_cal_one
from ..._fiff.meas_info import create_info
from ...utils import logger, verbose, fill_doc, _soft_import
```

File: mne/io/neuralynx/neuralynx.py [hunk 3]
```diff
@@ -0,0 +1,85 @@
...
+
+
+@fill_doc
+class RawNeuralynx(BaseRaw):
+    @verbose
+    def __init__(self, fname, *, preload=False, verbose=None):
+        try:
+            from neo.io import NeuralynxIO
+        except Exception:
+            raise ImportError("Missing the neo-python package") from None
```
Comment (larsoner) [inline mne/io/neuralynx/neuralynx.py:39]: We have helpers for a lot of common patterns like this one
```suggestion
        neo_io = _soft_import("neo.io", "reading Neuralynx data", strict=True)
        NeuralynxIO = neo.io.NeuralynxIO
```

File: mne/io/neuralynx/neuralynx.py [hunk 4]
```diff
@@ -0,0 +1,85 @@
...
+
+        datadir = op.abspath(fname)
```
Comment (larsoner) [inline mne/io/neuralynx/neuralynx.py:41]: another one
```suggestion
        datadir = _check_fname(fname, "read", True, "fname", need_dir=True)
```

File: mne/io/neuralynx/neuralynx.py [hunk 5]
```diff
@@ -0,0 +1,85 @@
...
+        block_id = 0  # assumes there's only one block of recording
+        n_total_samples = sum(
+            [
+                nlx_reader.get_signal_size(block_id, segment)
+                for segment in range(n_segments)
+            ]
+        )
+
+        # loop over found filenames and collect names and store last sample numbers
+        last_samps = []
```
Comment (larsoner) [inline mne/io/neuralynx/neuralynx.py:66]:
```suggestion
        last_samps = [n_total_samples - 1]
```

File: mne/io/neuralynx/neuralynx.py [hunk 6]
```diff
@@ -0,0 +1,85 @@
...
+            ncs_fname = nlx_reader.ncs_filenames[chan_key]
+            ncs_fnames.append(ncs_fname)
+            last_samps.append(
+                n_total_samples - 1
+            )  # assumes the same sample size for all files/channels
+
+        super(RawNeuralynx, self).__init__(
+            info=info,
+            last_samps=last_samps,
+            filenames=ncs_fnames,
```
Comment (larsoner) [inline mne/io/neuralynx/neuralynx.py:79]: Just store the dirname in `filenames` (it's really just used for `__repr__`), no need to store the other filenames I don't think
```suggestion
            filenames=[fname],
```

File: mne/io/neuralynx/tests/test_neuralynx.py [hunk 7]
```diff
@@ -0,0 +1,5 @@
+from neuralynx import read_raw_neuralynx
+from neuralynx import RawNeuralynx
+
+test_data = ""
+raw = RawNeuralynx(fname=test_data)
```
Comment (larsoner) [inline mne/io/neuralynx/tests/test_neuralynx.py:5]: See how other `mne/io` tests work, e.g. `mne/io/nsx/tests/test_nsx.py`

File: mne/io/neuralynx/neuralynx.py [hunk 8]
```diff
@@ -0,0 +1,85 @@
...
+            preload=preload,
+        )
+
+    def _read_segment_file(self, data, idx, fi, start, stop, cals, mult):
+        """Read a chunk of raw data."""
+        _read_segments_file(self, data, idx, fi, start, stop, cals, mult, dtype="<i2")
```
Comment (larsoner) [inline mne/io/neuralynx/neuralynx.py:85]: I'd expect this to be something like:
```suggestion
        nlx_reader = NeuralynxIO(dirname=fname)
        # no idea what method you need here, but somehow
        # get "block" to be filled with data from all channels
        # over the span `start:stop` (in samples), i.e.,
        # have shape (n_channels, stop - start)
        block = nlx_reader.get_data(start, stop)
        _mult_cal_one(data, block, idx, cals, mult)
```
But if `_read_segments_file` *actually* already works here (somehow), then you could loop over the filenames (after storing them in `_raw_extras`)

**Comment by larsoner** [comment]
Looks like a good start! I left some initial comments. Let me know if they don't make sense and I can clarify further. Or if you want me to try pushing some commits I can.

A good next step would be to get ~1 sec of raw data (can be junk!) added to mne-testing-data:

https://github.com/mne-tools/mne-testing-data#add-or-change-files-in-the-repo-for-use-with-mne

File: mne/io/neuralynx/neuralynx.py (hunk 3 above)
Comment (drammock) [reply mne/io/neuralynx/neuralynx.py:39]: last line should be

```py
NeuralynxIO = neo_io.NeuralynxIO
```
(neo_io underscore, not neo.io dot)

**Comment by KristijanArmeni** [comment]
Thanks for the detailed comments! All makes sense to me! I'll have a look and patch these in. In the meantime, I've asked around in the lab if there's a way to get the tiny sample recording done. Report back sometime next week once I hear back.

**Comment by KristijanArmeni** [comment]
Just pushed an update to my remote branch. Seeing the commit history chaos, I should prob stick to rebasing mne:main into my local development branch as opposed to merging it.

We now have the [testing dataset](https://github.com/mne-tools/mne-testing-data/tree/master/neuralynx) (dummy recording) available and I am trying to implement `mne.io.read_raw_neuralynx(neuralynx_testing_data)` (via [`RawNeuralynx()`](https://github.com/KristijanArmeni/mne-python/blob/b7279f38b76bbee495e98319f33c8002a8799c39/mne/io/neuralynx/neuralynx.py#L33) and `_read_segments_file()`) and plot the outcome. Currently, it will obtain the right number of channels and samples, but the values appear faulty (and are the same across channels). See below for details on how I benchmark.

I don't have a good understanding of `_read_segments_file()` yet, but I guess something goes wrong before `np.fromfile` is called to obtain the file (`.ncs`) contents. Perhaps I am doing something wrong in my call to [_read_segments_file](https://github.com/KristijanArmeni/mne-python/blob/b7279f38b76bbee495e98319f33c8002a8799c39/mne/io/neuralynx/neuralynx.py#L83)?

### Values read by read_raw_neuralynx() don't match those obtained by neo.NeuralynxIO().read() 

As a first attempt, I tried using `_read_segments_file` and setting `n_channels` argument to `1` (in [neuralynx.py](https://github.com/KristijanArmeni/mne-python/blob/b7279f38b76bbee495e98319f33c8002a8799c39/mne/io/neuralynx/neuralynx.py#L83)), because every `.ncs` file on disk is 1 channel. This does make the code run, but the outcome is faulty. For now, I benchmark against the `neo.neuralynxIO()` reader ([test_neuralynx.py](https://github.com/KristijanArmeni/mne-python/blob/nlx/mne/io/neuralynx/tests/test_neuralynx.py), see figure below for 3 channels):
- the number of samples/timepoints and channels matches matches well
- the actual sample values are different and most samples obtained by `_read_segments_file()` are actually 0 (not shown in figure).
- the values obtained by neo seem to make more sense to me (for an unconnected recording)

test_samples

**Comment by larsoner** [comment]
Wasn't sure the easiest way to fix the history so I did:
```
git checkout main
git fetch upstream
git reset --hard upstream/main
git checkout KristijanArmeni/nlx mne/io/neuralynx
# hack in a few tweaks
git commit --amend -a
# modify commit message
git push origin -u nlx
```
This gave the commit credit to you but created this branch / commit:

- https://github.com/larsoner/mne-python/tree/nlx
- https://github.com/larsoner/mne-python/commit/3ce7d8ba0e3d72345982c4dbe225745f0d72af82

If you think this commit is okay you can do locally:
```
git remote add larsoner git@github.com:larsoner/mne-python.git
git fetch larsoner
git checkout nlx
git checkout -b nlx-bak  # create a local backup!
git checkout nlx
git reset --hard larsoner/nlx
git commit --amend -a
git push origin --force nlx
```
and the one commit should be shown in this PR.

I'll look into the data structures soon

**Comment by larsoner** [comment]
Okay I got tests to pass here (still on my `nsx` branch):

https://github.com/larsoner/mne-python/commit/5cb32a4910868982d602f624ab9b558c0e7c37f8

> (via RawNeuralynx() and _read_segments_file())

Keep in mind `_read_segments_file` is a very dumb / simple function -- it assumes data in the filename are stored starting at byte zero with no metadata. So it will only work if that's the case for Neuralynx files. Is it?

You can see in my commit I just use `neo` for reading the data currently. It uses a very inefficient algorithm -- read all data from disk then subselect -- that we should improve upon. But hopefully it makes conceptually clear how things should be structured, and what `self._read_segment_file` must do internally.

**Comment by KristijanArmeni** [comment]
Thanks for suggesting the fix for the commit history! Will try this in a bit and see how it goes. The fixes in the last commit on your branch all make sense to me.

- re: header info in `.ncs`. Right, good call, indeed each `.ncs` does have metadata I think, so there'd need to be a separate reader for header info, which is what probably causes the mess.
- re: neo reading: I used neo to have something to benchmark against. I'm fine using it for reading in the data and I'm almost certain it has a way of reading in data after samples and channels are specified. Will have a look once I get this up and running.

File: mne/io/neuralynx/neuralynx.py [hunk 9]
```diff
@@ -0,0 +1,159 @@
...
+                    channel_indexes=idx, time_slice=(time[0], time[-1])
+                ).magnitude
+                for seg, time in zip(
+                    bl[0].segments[first_seg : last_seg + 1], sel_times
+                )
+                for signal in seg.analogsignals
+            ]
+        ).T
+
+        block = all_data  # shape = (len(idx), n_samples))
```
Comment (larsoner) [inline mne/io/neuralynx/neuralynx.py:153]: Perhaps somewhat confusingly, whatever you pass to `_mult_cal_one` needs to be shape `(n_channels, n_samples)`, not `(len(idx), n_samples)`. I'll push a quick commit to show what I mean

**Comment by larsoner** [comment]
With my commit now I see:
```
mne/io/neuralynx/tests/test_neuralynx.py:43: in test_neuralynx
    _test_raw_reader(read_raw_neuralynx, fname=testing_path)
mne/io/tests/test_raw.py:168: in _test_raw_reader
    data2, times2 = other_raw[picks, sl_time]
mne/io/base.py:851: in __getitem__
    return self._getitem(item)
mne/io/base.py:858: in _getitem
    data = self._read_segment(start=start, stop=stop, sel=sel)
<decorator-gen-311>:12: in _read_segment
    ???
mne/io/base.py:479: in _read_segment
    _ReadSegmentFileProtector(self)._read_segment_file(
mne/io/base.py:2511: in _read_segment_file
    return self.__raw.__class__._read_segment_file(
mne/io/neuralynx/neuralynx.py:162: in _read_segment_file
    block[idx] = all_data  # shape = (n_channels, n_samples)
E   ValueError: shape mismatch: value array of shape (10,6783) could not be broadcast to indexing result of shape (10,2000)
```
So the `all_data` that you create does not have enough samples, so there must be some bug with `_find_first_last_segment` or how it's used

File: mne/io/neuralynx/neuralynx.py [hunk 10]
```diff
@@ -0,0 +1,165 @@
...
+            # gather the start and stop times (in sec) for all signals in segments
+            # (assumes 1 signal per segment)
+            # shape = (n_segments, 2) where 2nd dim is (start_time, stop_time)
+            onst_offt = np.array(
+                [
+                    (signal.t_start.item(), signal.t_stop.item())
+                    for segment in segments
+                    for signal in segment.analogsignals
+                ]
+            )
```
Comment (larsoner) [inline mne/io/neuralynx/neuralynx.py:105]: Does this mean that there can be gaps and/or overlaps?

For gaps there is a way to handle it (insert zeros) but for overlaps I'm not even sure what the right behavior would be

Comment (KristijanArmeni) [reply mne/io/neuralynx/neuralynx.py:105]: Oh, good point. I'd assume no gaps/overlap as it should just be chunked representation of a continuous recording by the acquisition system (for some reason). That is, each segments' off time should be the next segments' start time etc. Our test dataset only has two segments (over a total of ~3 sec recording). I can perhaps test this on my local datasets which have on the order of 10's of segments.

Comment (larsoner) [reply mne/io/neuralynx/neuralynx.py:105]: If it's assumed that they are contiguous then I'd just construct an array of shape `(n_samples,)` of `int` type that gives you the mapping from sample numbers into segment numbers in `__init__` and store it in `_raw_extras[0] = dict(segment_map=segment-map)` or so. Then use this to figure out which segments to read in `_read_segments_file`. If this isn't clear I can push a quick commit to show what I mean

Comment (KristijanArmeni) [reply mne/io/neuralynx/neuralynx.py:105]: Thanks! Makes total sense. Will implement.

**Comment by KristijanArmeni** [comment]
OK, added a `smp2seg` dict in `_raw_extras[0]` in [df66e3](https://github.com/mne-tools/mne-python/pull/11969/commits/df66e338fba29ba8fa5e71cfbe386b0ac92b98c0).

I think this will work as a strategy to keep track of which samples belong to which segments. Seems like atm the tests are failing, but I can't see exactly where from the bot messages on here (stylistic things?).

In any case, if I manually evaluate `test_neuralynx()` locally it now runs with no errors. But if I try reading one of my bigger datasets I have locally I get some inconsistent sample numbers between `all_data` and `block` variables `RawNeuralynx._read_segment_file()`.

I suspect something fishy is going on with chunk reading based on time slices via`AnalogSignal.load(time_slice=(start_time, stop_time))`. It seems the concatenated arrays it loads based on `(start_time, stop_time)` don't have the same number of samples as an array which is constructed from `np.zeros(n_channels, stop-start)`.

Will explore and report here.

**Comment by KristijanArmeni** [comment]
In e251d98 I implemented way to read in segments via `neo` based on sample and channel indices (`Segment.AnalogSignal.load(channel_indexes=idx).magnitude[:, segment_start:segment_stop]`) as opposed to converting time/stop to time units and using `t_start/t_stop` in `AnalogSignal.load()` (this, for some reason, resulted in array inconsistencies on larger datasets). This code still only loads in memory the data segments between `start` and `stop` and channels specified in `idx`. See figures below for extra info.

**Filtering by channel type needed in read_raw_neuralynx()**.
After testing my implementation on local datasets that have several GBs worth of real data (and multiple channel types, e.g. `seeg`, `eog` etc.), I learned that channel types can have different number of segments which causes `NeuralynxIO.read()` to throw (a meaningful) error. `NeuralynxIO()` has the `exclude_filename` param which is a list of files to ignore to deal with this, such that the user can initiate separate readers for seeg, eog etc. (see [this github issue](https://github.com/NeuralEnsemble/python-neo/issues/1292)).

For now, my hack is to include `exclude_fnames_pattern` param containing file patterns to ignore (e.g. `"EOG*"`) in init. These patterns are passed to `glob.glob` and the returned lists of files are stored in `raw._extras` and passed to to `NeuralynxIO()` as `exclude_filename` input arg. Not sure if there’s a way of doing this that’s more aligned with the MNE API, but this worked for me locally.

### Extra information: raw.plot() on testing dataset

```python
from mne.datasets.testing import data_path, requires_testing_data
from mne.io import read_raw_neuralynx
testing_path = data_path(download=False) / "neuralynx"
raw = read_raw_neuralynx(testing_path, preload=True)
raw.plot()
```
test_raw_plot1


#### And an actual (non-dummy) dataset:  

This is a larger dataset (about 5GB in size).  
test_raw_plot2

File: mne/io/neuralynx/neuralynx.py [hunk 11]
```diff
@@ -0,0 +1,187 @@
...
+
+from ..._fiff.meas_info import create_info
+from ..._fiff.utils import _mult_cal_one
+from ...utils import _check_fname, _soft_import, fill_doc, logger, verbose
+from ..base import BaseRaw
+
+
+@fill_doc
+def read_raw_neuralynx(
+    fname, preload=False, verbose=None, exclude_fname_patterns: list = None
```
Comment (larsoner) [inline mne/io/neuralynx/neuralynx.py:14]: Let's add type hints later. Also `verbose` should always be last
```suggestion
    fname, *, preload=False, exclude_fname_patterns=None, verbose=None
```

File: mne/io/neuralynx/neuralynx.py [hunk 12]
```diff
@@ -0,0 +1,187 @@
...
+):
+    """Reader for Neuralynx files.
+
+    Parameters
+    ----------
+    fname : path-like
+        Path to a folder with Neuralynx .ncs files
+    %(preload)s
```
Comment (larsoner) [inline mne/io/neuralynx/neuralynx.py:22]:
```suggestion
    %(preload)s
    exclude_fname_patterns : list of str
        List of glob-like string patterns to exclude from channel list.
        Useful when not all channels have the same number of samples
        so you can read separate instances.
```

File: mne/io/neuralynx/neuralynx.py [hunk 13]
```diff
@@ -0,0 +1,187 @@
...
+@fill_doc
+def read_raw_neuralynx(
+    fname, preload=False, verbose=None, exclude_fname_patterns: list = None
+):
+    """Reader for Neuralynx files.
+
+    Parameters
+    ----------
+    fname : path-like
+        Path to a folder with Neuralynx .ncs files
```
Comment (larsoner) [inline mne/io/neuralynx/neuralynx.py:21]:
```suggestion
        Path to a folder with Neuralynx .ncs files.
```

**Review by larsoner — COMMENTED** [review]
Can you add to the test some use of `exclude_fname_patterns`? For example exclude some file or files and assert that those channels are not in the returned instance would probably be good enough.

Any chance you could record a tiny test dataset with different numbers of samples to actually test `exclude_fname_patterns` *and* raising a nice error on `read_raw_neuralynx` when reading a file where not all (non-excluded) channels have the same number of samples? If not we can probably live without it but it would be nice to have if it's easy to record.

**Comment by larsoner** [comment]
Oh and we'll also need a `doc/changes/devel.rst` update to mention this improvement!

**Comment by KristijanArmeni** [comment]
> Can you add to the test some use of `exclude_fname_patterns`? For example exclude some file or files and assert that those channels are not in the returned instance would probably be good enough.

Sure! W.r.t `exclude_fname_patterns`: alternative API design consideration could also be to shift the responsibility of file globing/filtering/selection to the user and expect the user to provide a correct list of files to be excluded. This would exactly mirror the `neo.NeuralynxIO()` api which also expects a list of files to exclude, so we would just pass the argument along essentially. The reason I am thinking of this is because if a user wants to, say, only _include_ EOG* channels, then they'd need to filter out and pattern match all `seeg` channels (e.g. `[LAIN*, RAIN*, LAM*, RAM*, ...]`), but not sure how consistent these labels are across datasets meaning globing might not always be the right solution. If left to the users, then they have the flexibility to hard-code the list, pattern match or whatever and pass the resulting list along.

> Any chance you could record a tiny test dataset with different numbers of samples to actually test `exclude_fname_patterns` _and_ raising a nice error on `read_raw_neuralynx` when reading a file where not all (non-excluded) channels have the same number of samples? If not we can probably live without it but it would be nice to have if it's easy to record.

Yeah, I can definitely ask! It's not difficult to record I think. Not sure how different sample sizes happen on the acquisition side to begin with, but I can ask if they can hook an EOG channel to the setup.

I'll also ask them to provide a `.mat` file of the testing dataset (afaik neuralynx ships with a neuralynx2mat.m script) such that we can cross-check `assert_allclose(raw.get_data(), testing_dataset.mat)`?

**Comment by larsoner** [comment]
> alternative API design consideration could also be to shift the responsibility of file globing/filtering/selection to the user and expect the user to provide a correct list of files to be excluded

I don't mind the list of glob. Users can still pass explicit filenames if you want so it just makes things easier sometimes / hopefully.

> I'll also ask them to provide a .mat file of the testing dataset (afaik neuralynx ships with a neuralynx2mat.m script) such that we can cross-check assert_allclose(raw.get_data(), testing_dataset.mat)?

That would be great. If you could also run this on the existing `mne-testing-data` dataset and add the `.mat` file there, it would provide better validation than using `neo` explicitly in the test (though we could keep that anyway just to make sure 1) RawNeuralinx, 2) using neo directly, and 3) neuralynx2mat all agree on data values.

**Comment by KristijanArmeni** [comment]
OK, added all commits that I have locally in this branch. At least having these changes + the new version of mne-testing-data locally let's me run `test_neuralynx()` fine. I see some tests are still not passing, might be cuz I'm not running some style formatting on the code before commiting?

But basically, the last four commits:
- `test_neuralynx()` now it loads .mat files via `scipy.io.loadmat` (had to write a small reader function on top of that)
- it tests that `exclude_fname_patterns` works in `read_raw_neuralynx()`
- there's a more informative error to warn the users if channels (i.e. .ncs files) have different number of sample points (which is detected by `NeuralynxIO`)
- it tests that data sample outputs are all close in a three way fashion: `read_raw_neuralynx()` vs. `neo.NeuralynxIO` vs. `Nlx2Mat.m`

**Comment by larsoner** [comment]
> might be cuz I'm not running some style formatting on the code before commiting?

Yes if you click blue "details" to the right of the failing `Tests/Style` to get to https://github.com/mne-tools/mne-python/actions/runs/6852318442/job/18630579970?pr=11969 you'll see:
```
mne/io/neuralynx/tests/test_neuralynx.py:18:5: D205 1 blank line required between summary line and description
mne/io/neuralynx/tests/test_neuralynx.py:18:5: D400 First line should end with a period
mne/io/neuralynx/tests/test_neuralynx.py:18:5: D401 First line of docstring should be in imperative mood: "Helper function to convert the read-in "Header" field into a dict."
mne/io/neuralynx/tests/test_neuralynx.py:35:5: D103 Missing docstring in public function
```

**Comment by larsoner** [comment]
> can't tell if a pushed commit alone also sends a notification to all involved in the conversation

It can, but most maintainers I know have disabled this feature because it generates way too many emails, so yes please do comment when it's time to look. Failure is unrelated so you can ignore, I'll merge `main` into this branch to fix it. Can you take out of draft status if you're ready for it to be merged (assuming CIs are green)? Then I'll take one last look and merge if it's green and I have no further comments!

**Comment by KristijanArmeni** [comment]
@larsoner It looks all green? See if there's anything missing or fishy in the reader.

It should be a minimally working reader. I guess other improvements can follow on top of this, like bookkeeping any additional info from nlx headers (meas_date etc.). I'll be using this on real data in the coming weeks, so I can open issues as they pop along.

> Oh and we'll also need a doc/changes/devel.rst update to mention this improvement!

This can be done in a separate PR? Once it's merged, I'd like to also acknowledge the researcher who acquired the testing dataset, provided the `.mat` files and helped troubleshooting Neuralynx specifics (if that's a practice).

**Comment by larsoner** [comment]
> This can be done in a separate PR?

It should be part of this one if possible. Feel free to push.

Yes you can credit whoever you think should get co-author credit for the PR by adding names at the end of the changelog line (or multiple `:newcontrib:`s for people who are new contributors).

**Comment by KristijanArmeni** [comment]
Thanks @larsoner! I updated `doc/changes/devel.rst` and `names.ini`, hopefully in good order. Seems all checks passing.

**Comment by larsoner** [comment]
Thanks @KristijanArmeni !

**Comment by eduardosand** [comment]
@KristijanArmeni This is awesome! I was wondering if the end how you deal with missing samples, i.e. array size mismatch(when making an array using first and last timepoints vs the actual data present). For me, having to use the exclude file parameter seems like it disregards the power of having all the data in the directory at once and a warning is nice, but not if there isn't much I can do about it.

On my end, I have two opinions on the matter.

The sample rate itself is not exactly 32K or whatever ncs_reader.get_signal_sampling_rate() spits out, but something slightly smaller than that. This could contribute to the reason why there is a slight mismatch, and why this increases on larger datasets.

Segments are generated in neo, not just when the experiments presses play/pause or an event annotation is made, but also when neuralynx fails to record certain samples. I also think this is the reason for the difference in samples across channels, and as the github issue above points out, there's also some weirdness with cutting off samples at the beginning and end of a file. Hence, neuralynx will drop more samples in the 32K sampled channels than the 8K sampled channels but asymmetrically(ex. neuralynx drops 3-4 samples in 32K case, no samples are dropped in 8K case, so a segment is generated in 32K channels, but not 8K channels).

See below for more:
https://www.fieldtriptoolbox.org/faq/discontinuous_neuralynx/
https://github.com/fieldtrip/fieldtrip/issues/1128


My workaround has thus been to assume uniform sampling within a segment, and then interpolating with cubic splines for any data 'in between' segments. I do this instead of fieldtrips filling with NaNs, because the signal processing is easier downstream (without having to then recast the NaN values to something suitable for scipy.signal.sosfiltfilt). Outside of spike sorting, I don't care about 32K or 8K sampling rate, so I then downsample to 2K and save as .npy files for downstream analysis.

**Comment by larsoner** [comment]
> For me, having to use the exclude file parameter seems like it disregards the power of having all the data in the directory at once and a warning is nice, but not if there isn't much I can do about it.

Assuming you have two streams to align / resample (though it generalizes to more), naively my approach would be to use the `exclude` parameter to create two `Raw` instances (the first instance excluding the second file, and the second instance excluding the first file) and then resample them.

If we have gaps in some recordings we should make sure these are in there properly with zeros (or NaNs) and annotated with `BAD_ACQ_SKIP` so that time is continuous in the instance. Then we "just" need to work out https://github.com/mne-tools/mne-python/pull/11408 to resample the data to the same sample rate.

**Comment by eduardosand** [comment]
> If we have gaps in some recordings we should make sure these are in there properly with zeros (or NaNs) and annotated with `BAD_ACQ_SKIP` so that time is continuous in the instance. Then we "just" need to work out #11408 to resample the data to the same sample rate.

I agree with annotation, but I was under the impression that scipy.signal doesn't work well with NaNs? Also worried that 0s might cause edge effects when computing power analyses, although maybe not as big of a concern with wavelets?
Interpolation is nice because it allows us to use any filters we want and as long as remember the values themselves are interpolated, we can discard them later after pre-processing.

**Comment by larsoner** [comment]
> I agree with annotation, but I was under the impression that scipy.signal doesn't work well with NaNs? Also worried that 0s might cause edge effects when computing power analyses, although maybe not as big of a concern with wavelets?

Correct, it doesn't handle NaN well -- it will turn the whole signal `NaN`. But the idea isn't to operate on the whole signal at once. See [the original issue](https://github.com/mne-tools/mne-python/issues/10447#issue-1174683698) -- the idea is to resample each non-NaN segment separately using `skip_by_annotation` to find segments that should be resampled. Similarly we can improve our `tfr_*` functions to also respect `skip_by_annotation` and thereby (again) process each segment separately.

But thinking about it more, getting the segment-by-segment resampling right seems like a difficult problem. Also reading through your description of Neuralynx and samples just missing at the beginning and end of a recording (instead of in the middle someplaces) I'm not totally sure you need all this. So maybe not the right option here 🤷

**Comment by KristijanArmeni** [comment]
Thanks for raising this @eduardosand! You mention two sources/forms of array-size mismatch: 

1. Across channels due to different sampling rates and/or asymmetric Neuralynx sample trimming (e.g. 8kHz, vs 32KHz, this I was aware of this is what, as @larsoner said, `exclude_fname_pattern`, propagated to `neo` reader, should take care of such that the user can construct two Raw instances if need be; outside of whether or not the two can eventually be realigned). 
2. Gaps across segments, discontinuous recording within a channel (i.e. number of valid samples is smaller than total number of samples). Indeed, the current code assumes the segments are continuous, temporally adjacent.

Regarding 2 and reading in the Fieldtrip doc you linked:

> The low-level `read_neuralynx_ncs` function detects the presence of gaps in the `.ncs` file and issues a warning.

It sounds like `mne.io.read_raw_neuralynx()` should minimally check for temporal gaps between `neo` segments and issue a warning? For example, it should check that each `neo.Segment[i]` object starts when the `neo.Segment[i-1]` ended and raise a warning if this is not the case (i.e. there's temporal gap, assuming the information in `neo` is accurate)? And potentially also reconstruct/fill/mark missing samples such that the time axis (i.e. `raw.times`) is continuous and valid.

If this is on track, happy to open a separate issue and work on this.

**Comment by larsoner** [comment]
Sounds good, let's move discussion to #12247

**Pull request merged**